### PR TITLE
Issue 3173: Prevent mis-routing of events

### LINKED ISSUE #3173: Non-unique projection of client key to segmentstore key
**Problem description**
We are running a longevity test where the client writes to a stream at a constant event rate to 20 randomly pre-chosen keys. The stream has an event-rate triggered scaling policy. There are three writers in one writer group.

The event rate for an individual client key is **lower** than the scaling trigger, and we expect that for each value of the 20 client keys it would be possible to have a segment of a **finite** width that contains only one key, so we don't expect any segments to be splitting infinitely.

However, in reality we observe that after one day of running the test the stream arrives at a state when the number of  stream segments is less than 20 (the number of client keys) and almost half of the segments are empty (not being written into).
As to the active segments, some of them are infinitely narrow within the numerical precision (so that further splitting is impossible) and have the write rates that are double of the client write rate. 
This is the indication that more than one String-typed client key is being written to the same hash value of the segment key.

Example of the stream that stopped scaling at 19 segments per stream:
heat-chart
(most of the active segments here are very narrow to be seen and are located between the light-blue inactive segments).

The client is writing to the following 20 keys:
```
0f5c8e82-4a90-4dab-b866-745cac949d72
1203a27d-d5aa-405c-87ba-cc4811b80e82
22cd4044-32a3-4f59-b49b-c467bdc09aaf
37138de7-c701-4042-bc9d-91a1fc52d4be
41f79861-39fe-4231-a980-d84e8a753c5d
452e4ea9-ca70-48a8-9f4b-59a0c7d3030f
4a83d604-83ca-4e27-9119-48d764ef4f2b
7eb437db-631d-49e8-8fd6-46c5bda879d1
8133ce55-0206-4e62-9991-c4c89e4d5822
883e76c3-6556-4751-93d8-08e9a734ab9d
93c9f3ad-a76e-4046-9fe6-63909c303988
aab63967-6ca0-4b15-89f3-e0625a558e54
ad88ceb4-703a-4fc0-9102-4ab6f4d6fab4
b476f659-e8f3-42b1-bea4-47c691c3a6fe
cf45b9fc-3244-4ec8-a26e-298a52866a1c
d6a890a4-051d-4f63-8b22-dade01cc0a6d
f617e1c4-eb5a-4e85-b576-1f0e5c9466bc
f62a2f91-7fe2-4e49-aa54-0b0c1cee4cdf
e7851825-f9b6-4601-bd42-0ce4fad2fb26
e94ce3e6-26d4-48ab-a48a-448aa9bd5203
```
(See the [client log file](https://github.com/pravega/pravega/files/2650036/logs-from-small-scale-young-worker-in-small-scale-young-worker-6fcd78db45-r99d7.txt.zip) for detail)

However, only 11 segments are being written into:
write_rates 

Looking at the response from the `scaling_events` API, 9 of the 11 active segments  (namely 904, 991, 1259, 1346, 1363, 1411, 1584, 1638, and 1656) have zero width (to the numerical precision) and have the write rates of 3 kB/s which is twice the 1.5 kB/s write rate per single client key. Check the [attached file](https://github.com/pravega/pravega/files/2650043/scaling.txt) for the exact key values.
So each of this 9 segments has its numerical segment key value being mapped from two distinct string values of client key.  

The two remaining active segments (1391 and 1647) have finite width and a write rate of 1.5 kB/s, suggesting that they are getting writes from one client key each.

The above behavior is observed for commit `26fb466`. 
It did not happen when we were running `0.3` pravega.

**Problem location**
Segmentstore

**Suggestions for an improvement**
Client keys should be uniquely mapped to segment keys

### PR DESCRIPTION
**Pull request by tkaitchuck**
**Change log description**  
* Prevents a bug in a rare edge case where events from a writer would be consistently sent to the 'incorrect' segment because the internal map was incorrect.

**Purpose of the change**  
Hopefully fixes #3173

**What the code does**  
In the case where a segment was sealed and replaced unevenly (not a simple split or merge) and the uneven segments were sealed and replaced before any data was written to them, but and the writer saw the successor to the successor segment before one of the segments from the original generation was sealed AND the controller returned the replacement segments in a non-ascending order, then the writer would alter the end range for the final segment effectively expanding it's keyspace. 

This could lead to bad load-balancing or the inability for splitting or merging to correctly manage the load.

Prior to this patch if a writer is in this state, simply restarting it will correct the problem. 

This won't explicitly violate our ordering guarantees because while events would be sent to the 'wrong' segment it would be done consistently. (On a cross writer basis the same key would end up on different segments, which is unusual, but not something we explicitly guarantee doesn't happen at the moment.)

**How to verify it**  
We need to run again with the same client in #3173 and see if the events are routed appropriately.

### CODE REVIEW

File: client/src/main/java/io/pravega/client/stream/impl/StreamSegments.java [hunk 1]
```diff
@@ -76,7 +74,8 @@ public StreamSegments withReplacementRange(Segment replacedSegment, StreamSegmen
         verifyReplacementRange(replacementRanges);
         NavigableMap<Double, Segment> result = new TreeMap<>();
         Map<Long, List<SegmentWithRange>> replacedRanges = replacementRanges.getReplacementRanges();
-        Optional<List<SegmentWithRange>> replacements = Optional.ofNullable(replacedRanges.get(replacedSegment.getSegmentId()));
+        List<SegmentWithRange> replacements = replacedRanges.get(replacedSegment.getSegmentId());
+        replacements.sort((a, b) -> Double.compare(b.getHigh(), a.getHigh()));
```
[the remark this answers is not in the document]
Comment (tkaitchuck) [reply client/src/main/java/io/pravega/client/stream/impl/StreamSegments.java]: It actually can't be null. If it were `verifyReplacementRange`4 lines up would throw an NPE.

[the remark this answers is not in the document]
Comment (andreipaduroiu) [reply client/src/main/java/io/pravega/client/stream/impl/StreamSegments.java]: Can this be replaced with `.sort(Comparators.comparingDouble(SegmentWithRange::getHigh))` ? Makes the syntax look nicer.

[the remark this answers is not in the document]
Comment (tkaitchuck) [reply client/src/main/java/io/pravega/client/stream/impl/StreamSegments.java]: >Can this be replaced with replacements.sort(Comparator.comparingDouble(SegmentWithRange::getHigh)) ?

I need a reverse sort. Is there an easy way to do that?

[the remark this answers is not in the document]
Comment (andreipaduroiu) [reply client/src/main/java/io/pravega/client/stream/impl/StreamSegments.java]: @tkaitchuck oops I didn't see that. There's no nice-looking syntax for that. You can tag a `.reverse()` onto the resulting object, but that won't make it look so nice:
```
replacements.sort(Comparator.comparingDouble(SegmentWithRange::getHigh).reverse())
```

**Pull request merged**